### PR TITLE
CCD-20613 - Compute API

### PR DESCRIPTION
**Pull request by jamesoleary123**
New Compute API to use List Historical Server Configurations for a Server

### CODE REVIEW

File: ComputeClient/Compute.Client.UnitTests/ApiUriTests.cs [hunk 1]
```diff
@@ -15,15 +15,15 @@ public void ReturnsGetDomainNatRulesUri()
         {
             var networkDomainId = Guid.NewGuid();
             var uri = ApiUris.GetDomainNatRules(_orgId, networkDomainId.ToString());
-            Assert.AreEqual(ApiUris.MCP2_9_PREFIX + _orgId + "/network/natRule?networkDomainId="+networkDomainId, uri.OriginalString);
+            Assert.AreEqual(ApiUris.MCP2_9_PREFIX + _orgId + "/network/natRule?networkDomainId=" + networkDomainId, uri.OriginalString);
```
Comment (jamesoleary123) [inline ComputeClient/Compute.Client.UnitTests/ApiUriTests.cs:18]: formated file using Visual Studio 2017 - real changes in this file start on line 158

File: ComputeClient/Compute.Client.UnitTests/Server20/ServerAccessorTests.cs [hunk 2]
```diff
@@ -10,55 +10,55 @@
...
 {
-	[TestClass]
-	public class ServerAccessorTests : BaseApiClientTestFixture
-	{
-		[TestMethod]
-		public async Task GetServers_ReturnsResponse()
-		{
-			requestsAndResponses.Add(ApiUris.MyAccount,RequestFileResponseType.AsGoodResponse("GetMyAccountDetails.xml"));
-			requestsAndResponses.Add(ApiUris.GetMcp2Servers(accountId), RequestFileResponseType.AsGoodResponse("GetServersResponse.xml"));
+    [TestClass]
```
Comment (jamesoleary123) [inline ComputeClient/Compute.Client.UnitTests/Server20/ServerAccessorTests.cs:13]: formated file using Visual Studio 2017 - real changes in this file start on line 266

File: ComputeClient/Compute.Client/ApiUris.cs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ﻿using System.Data;
```
Comment (jamesoleary123) [inline ComputeClient/Compute.Client/ApiUris.cs]: formated file using Visual Studio 2017 - real changes in this file start on line 3039

File: ComputeClient/Compute.Client/ApiUris.cs [hunk 4]
```diff
@@ -1,5 +1,5 @@
...
-using System.Text; 
+using System.Text;
```
Comment (jamesoleary123) [inline ComputeClient/Compute.Client/ApiUris.cs:3]: formated file using Visual Studio 2017 - real changes in this file start on line 3039

File: ComputeClient/Compute.Client.UnitTests/Server20/ServerAccessorTests.cs [hunk 5]
```diff
@@ -262,5 +262,123 @@ public async Task NotifyNicIpChange_ReturnsResponse()
             Assert.AreEqual("NOTIFY_NIC_IP_ADDRESS_CHANGE", response.operation);
             Assert.AreEqual("IN_PROGRESS", response.responseCode);
         }
+
+        [TestMethod]
+        public async Task ListHistoricalServerConfigurationsPaginated_ReturnsResponse()
```
Comment (jamesoleary123) [inline ComputeClient/Compute.Client.UnitTests/Server20/ServerAccessorTests.cs:267]: Test used to verify that sample API response payload contained in **ListHistoricalServerConfigurationResponse.xml** is correctly unmarshalled to _HistoricalServerConfigurations_

File: ComputeClient/Compute.Client/Server20/ServerAccessor.cs [hunk 6]
```diff
@@ -635,5 +635,39 @@ public async Task<ResponseType> CreateDrsTargetServer(CreateDrsTargetServerType
 		{
 			return await _apiClient.PostAsync<CreateDrsTargetServerType, ResponseType>(ApiUris.CreateDrsTargetServer(_apiClient.OrganizationId), createDrsTargetServer);
 		}
+
+        /// <summary>The list historical server configurations for mcp 2 deployed servers.</summary>
+        /// <param name="serverId">The server Id</param>
```
Comment (jimantz) [inline ComputeClient/Compute.Client/Server20/ServerAccessor.cs:640]: Do we need the "time" param in here too? It's a timestamp in the API Definition. We might need to check the format used etc. Not sure how timestamps and C# work.

File: ComputeClient/Compute.Client/Server20/ServerAccessor.cs [hunk 7]
```diff
@@ -635,5 +635,39 @@ public async Task<ResponseType> CreateDrsTargetServer(CreateDrsTargetServerType
...
+        /// <param name="filteringOptions">The filtering options.</param>
+        /// <returns>The <see cref="Task"/>.</returns>
+        public async Task <IEnumerable<HistoricalServerConfigurationType>> ListHistoricalServerConfigurations(Guid serverId, HistoricalServerConfigurationListOptions filteringOptions = null) {
+            var response = await ListHistoricalServerConfigurationsPaginated(serverId, filteringOptions, null);
```
Comment (jimantz) [inline ComputeClient/Compute.Client/Server20/ServerAccessor.cs:644]: Do we need the check for empty ServerID here, or can we let the API take care of that?

File: ComputeClient/Compute.Contracts/Server/HistoricalServerConfigurationType.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            }
+            set
+            {
+                this.startTimeField = value;
+            }
+        }
+
+        /// <remarks/>
+        [System.Xml.Serialization.XmlAttributeAttribute()]
+        public System.DateTime endTime
```
Comment (jimantz) [inline ComputeClient/Compute.Contracts/Server/HistoricalServerConfigurationType.cs:83]: Reading this: https://social.msdn.microsoft.com/Forums/en-US/ae260f91-2907-4f31-a554-74c8162b3b38/xsdexe-tool-creates-properties-with-quotspecifiedquot-postfix?forum=xmlandnetfx

Would changing to public System.DateTime? startTime allow us to remove the "specified" field?

File: ComputeClient/Compute.Client/Server20/ServerAccessor.cs (hunk 6 above)
Comment (jimantz) [reply ComputeClient/Compute.Client/Server20/ServerAccessor.cs:640]: Discussed - time param supplied in filters as it's not mandatory.

File: ComputeClient/Compute.Client/Server20/ServerAccessor.cs (hunk 7 above)
Comment (jimantz) [reply ComputeClient/Compute.Client/Server20/ServerAccessor.cs:644]: Call to "paginated" function will verify.

**Review by jimantz — APPROVED** [review]
Looks good, approved 👍

**Pull request merged**